diff --git i/src/bin/main.rs w/src/bin/main.rs
index 7f3f441..fdd3ffb 100644
--- i/src/bin/main.rs
+++ w/src/bin/main.rs
@@ -1,9 +1,9 @@
+use skip_list::parameter::Parameter;
 use skip_list::user::{build_idx, query_idx};
 use std::fs::File;
 use std::io;
 use std::io::{BufReader, Read};
 use std::path::{Path, PathBuf};
-use skip_list::parameter::Parameter;
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
     let args: Vec<String> = std::env::args().collect();
@@ -15,7 +15,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     let parameter = Parameter {
         factor_size: args[2].parse().unwrap_or(1),
-        factor_skip: args[3].parse().unwrap_or(1)
+        factor_skip: args[3].parse().unwrap_or(1),
     };
     let file_stem = &args[1];
 
@@ -43,7 +43,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
             let mut block_result = 0u64;
             for num in data[block_i].as_slice() {
                 if query == num {
-                    block_result +=  1;
+                    block_result += 1;
                 }
             }
             expected_result += block_result;
@@ -62,9 +62,18 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     }
     let total_index_size_kb = total_index_size_bytes / 1024;
     let max_score = parameter.factor_skip * queries.len() as i64 * data.len() as i64;
-    let score = (n_skipped as i64 * parameter.factor_skip) - (total_index_size_kb as i64 * parameter.factor_size);
-    println!("storage size: {} {}", total_index_size_kb, total_index_size_kb as i64 * parameter.factor_size);
-    println!("num skips: {} {}", n_skipped, n_skipped as i64 * parameter.factor_skip);
+    let score = (n_skipped as i64 * parameter.factor_skip)
+        - (total_index_size_kb as i64 * parameter.factor_size);
+    println!(
+        "storage size: {} {}",
+        total_index_size_kb,
+        total_index_size_kb as i64 * parameter.factor_size
+    );
+    println!(
+        "num skips: {} {}",
+        n_skipped,
+        n_skipped as i64 * parameter.factor_skip
+    );
     println!("total score: {}", score / max_score);
     Ok(())
 }
@@ -74,22 +83,21 @@ fn load_data(file_path: &Path) -> io::Result<Vec<Vec<i32>>> {
 
     let mut buf8 = [0u8; 8];
     file.read_exact(&mut buf8)?;
-    let n = u64::from_le_bytes(buf8) as usize;
+    // let n = u64::from_le_bytes(buf8) as usize;
 
-    file.read_exact(&mut buf8)?;
+    // First: number of blocks
     let n_block = u64::from_le_bytes(buf8) as usize;
+    file.read_exact(&mut buf8)?;
 
-    let chunk_size = (n + n_block - 1) / n_block;
+    // Second: values per block
+    let chunk_size = u64::from_le_bytes(buf8) as usize;
     let mut blocks: Vec<Vec<i32>> = Vec::with_capacity(n_block);
 
     let mut buf4 = [0u8; 4];
-    for block_idx in 0..n_block {
+    for _block_idx in 0..n_block {
         let mut chunk = Vec::with_capacity(chunk_size);
         for _ in 0..chunk_size {
             // stop exactly at n elements
-            if chunk.len() + block_idx * chunk_size >= n {
-                break;
-            }
             file.read_exact(&mut buf4)?;
             chunk.push(i32::from_le_bytes(buf4));
         }
@@ -114,4 +122,4 @@ fn load_query(file_path: &Path) -> io::Result<Vec<i32>> {
     }
 
     Ok(data)
-}
\ No newline at end of file
+}
